### ISSUE TITLE
`rustc` should prefer statically linking native dependencies with unspecified kind if `crt-static` is enabled

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#71647]
The main goal of `+crt-static` (at least on non-Windows targets) is to produce self-contained statically linked executables.

With this goal in mind `rustc` switches linking of 1) Rust crates and 2) libc to static if possible.
However, for native dependencies with unspecified kind `rustc` will still prefer the dynamic version of the library if both are available.

Unfortunately, doing this is not entirely trivial because `-Bstatic -lmylib` won't just prefer the static version (e.g. `libmylib.a`), it will *require* it.
So we need to check for the existence of `libmylib.a` first (in which directories exactly?) and then pass `-l:libmylib.a` instead of `-lmylib` if it exists.

cc https://github.com/rust-lang/rust/pull/71586

**Comment by petrochenkov** [comment]
>`-Bstatic -lmylib` won't just prefer the static version (e.g. `libmylib.a`), it will *require* it.

If https://github.com/rust-lang/rust/issues/39998 is accepted as the way to go by default (on non-Windows platforms), then we can just use `-Bstatic`.

**Comment by smaeul** [comment]
This was discussed at length in https://github.com/rust-lang/rust/pull/55566, where I was told that `NativeUnknown` means dynamic, and changing the semantics of `#[link]` was unacceptable.

**Comment by petrochenkov** [comment]
Some history in links: https://github.com/rust-lang/rust/pull/71586#issuecomment-626328134

**Comment by petrochenkov** [comment]
@smaeul 
I do not agree with what alexcrichton said https://github.com/rust-lang/rust/pull/55566, it's mostly a "it mostly works, so don't touch it because you may break something" position motivated by lack of resources and people ready to take responsibilities for this work on the Rust side.

We can *at least* make an experiment (possibly using crater, possibly landing this on nightly and reverting) and verify what actually breaks in practice.

**Comment by nagisa** [comment]
I fear crater will likely have  a pretty poor coverage as far as linkage is concerned as its environment won't generally have anything but the most common libraries installed on the system (and the "native" libraries built in `build.rs` are already almost universally statically linked).

**Comment by polarathene** [comment]
Was this not resolved with https://github.com/rust-lang/rust/pull/111698 ?
